spec/p2p/v0.34/pex.md

```diff
@@ -0,0 +1,111 @@
+and starts a routine that periodically persists the address book content to disk.
+
+The PEX reactor is configured with the address of a number of seed nodes,
+the `Seeds` parameter of the `ReactorConfig`.
+The addresses of seed nodes are parsed into `NetAddress` instances and resolved
+into IP addresses, which is implemented by the `checkSeeds` method.
+Valid seed node addresses are stored in the `seedAddrs` field,
+and are used by the `dialSeeds` method to contact the configured seed nodes.
+
+The last action is to start one of the following persistent routines, based on
+the `SeedMode` configuration parameter:
+
+- Regular nodes run the `ensurePeersRoutine` to check whether the node has
+  enough outbound peers, dialing peers when necessary
+- Seed nodes run the `crawlPeersRoutine` to periodically start a new round
+  of [crawling](./pex-protocol.md#Crawling-peers) to discover as much peer
+  addresses as possible
+
+### Errors
+
+Errors encountered when loading the address book from disk are returned,
+and prevent the reactor from being started.
+An exception is made for the `service.ErrAlreadyStarted` error, which is ignored.
+
+Errors encountered when parsing the configured addresses of seed nodes
+are returned and cause the reactor startup to fail.
+An exception is made for DNS resolution `ErrNetAddressLookup` errors,
+which are not deemed fatal and are only logged as invalid addresses.
+
+If none of the configured seed node adresses is valid, and the loaded address
+book is empty, the reactor is not started and an error is returned.
+
+## OnStop
+
+The `OnStop` method implements `BaseService` and stops the PEX reactor.
+
+The address book routine that periodically saves its content to disk is stopped.
+
+## GetChannels
+
+The `GetChannels` method, from the `Reactor` interface, returns the descriptor
+of the channel used by the PEX protocol.
+
+The channel ID is `PexChannel` (0), with priority `1`, send queue capacity of
+`10`, and maximum message size of `64000` bytes.
+
+## AddPeer
+
+The `AddPeer` method, from the `Reactor` interface,
+adds a new peer to the PEX protocol.
+
+If the new peer is an **inbound peer**, i.e., if the peer has dialed the node,
+the peer's address is [added to the address book](./addressbook.md#new-addresses).
+Since the peer was authenticated when establishing a secret connection with it,
+the source of the peer address is trusted, and its source is set by the peer itself.
+In the case of an outbound peer, the node should already have its address in
+the address book, as the switch has dialed the peer.
+
+If the peer is an **outbound peer**, i.e., if the node has dialed the peer,
+and the PEX protocol needs more addresses,
+the node [sends a PEX request](./pex-protocol.md#Requesting-Addresses) to the peer.
+The same is not done when inbound peers are added because they are deemed least
+trustworthy than outbound peers.
+
+## RemovePeer
+
+The `RemovePeer` method, from the `Reactor` interface,
```

Comment:
This function does not actually remove the peer from the address book, right?

Reply:
Nope. This is a method of the `Reactor` interface, which essentially means the peer was disconnected.